File: apps/api/app/schemas/agent.py
import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.models.agent import ExecutionStatus, OutputMode, TemplateCategory


class PromptTemplateRead(BaseModel):
    id: str
    slug: str
    name: str
    description: str | None = None
    category: str
    system_prompt: str
    user_prompt_template: str
    output_mode: str
    json_schema: dict[str, Any] | None = None
    tools: list[dict[str, Any]] | None = None
    default_provider: str
    default_model: str | None = None
    is_builtin: bool = False


class PromptTemplateCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    slug: str = Field(min_length=1, max_length=128)
    description: str | None = None
    category: TemplateCategory = TemplateCategory.CUSTOM
    system_prompt: str = Field(min_length=1)
    user_prompt_template: str = Field(min_length=1)
    output_mode: OutputMode = OutputMode.TEXT
    json_schema: dict[str, Any] | None = None
    tools: list[dict[str, Any]] | None = None
    default_provider: str = "openai"
    default_model: str | None = None


class AgentRunRequest(BaseModel):
    provider: Literal["openai", "anthropic"] | None = None
    model: str | None = None
    template: str | None = None
    system_prompt: str | None = None
    user_prompt: str | None = None
    variables: dict[str, Any] = Field(default_factory=dict)
    input: dict[str, Any] = Field(default_factory=dict)
    output_mode: OutputMode | str | None = None
    json_schema: dict[str, Any] | None = None
    tools: list[dict[str, Any]] | None = None
    workflow_id: uuid.UUID | None = None
    node_id: str | None = None
    temperature: float = Field(default=0.7, ge=0, le=2)
    max_tokens: int = Field(default=4096, ge=1, le=128000)


class WorkflowExecuteRequest(BaseModel):
    input: dict[str, Any] = Field(default_factory=dict)


class ExecutionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    workflow_id: uuid.UUID | None
    node_id: str | None
    template_slug: str | None
    provider: str
    model: str
    status: ExecutionStatus
    input_data: dict[str, Any]
    output_data: dict[str, Any] | None
    tool_calls: list[dict[str, Any]] | None
    steps: list[dict[str, Any]] | None
    usage: dict[str, Any] | None
    error_message: str | None
    duration_ms: int | None
    created_at: datetime
    completed_at: datetime | None


class WorkflowExecuteResponse(BaseModel):
    workflow_id: uuid.UUID
    executions: list[ExecutionRead]
    total_steps: int

File: apps/web/src/types/agent.ts
export type LLMProvider = "openai" | "anthropic";
export type OutputMode = "text" | "json";
export type ExecutionStatus = "pending" | "running" | "completed" | "failed";
export type TemplateCategory =
  | "lead_qualification"
  | "email_generation"
  | "meeting_summary"
  | "custom";

export interface PromptTemplate {
  id: string;
  slug: string;
  name: string;
  description: string | null;
  category: TemplateCategory;
  system_prompt: string;
  user_prompt_template: string;
  output_mode: OutputMode;
  json_schema: Record<string, unknown> | null;
  tools: Record<string, unknown>[] | null;
  default_provider: LLMProvider;
  default_model: string | null;
  is_builtin: boolean;
}

export interface AiNodeConfig {
  provider?: LLMProvider;
  model?: string;
  template?: string;
  systemPrompt?: string;
  userPrompt?: string;
  outputMode?: OutputMode;
  jsonSchema?: Record<string, unknown>;
  tools?: Record<string, unknown>[];
  temperature?: number;
}

export interface AgentRunPayload {
  provider?: LLMProvider;
  model?: string;
  template?: string;
  system_prompt?: string;
  user_prompt?: string;
  variables?: Record<string, unknown>;
  input?: Record<string, unknown>;
  output_mode?: OutputMode;
  json_schema?: Record<string, unknown>;
  tools?: Record<string, unknown>[];
  workflow_id?: string;
  node_id?: string;
  temperature?: number;
  max_tokens?: number;
}

export interface AgentExecution {
  id: string;
  user_id: string;
  workflow_id: string | null;
  node_id: string | null;
  template_slug: string | null;
  provider: string;
  model: string;
  status: ExecutionStatus;
  input_data: Record<string, unknown>;
  output_data: {
    content?: string;
    parsed?: Record<string, unknown>;
  } | null;
  tool_calls: {
    name: string;
    arguments: Record<string, unknown>;
    result: Record<string, unknown>;
  }[] | null;
  steps: Record<string, unknown>[] | null;
  usage: Record<string, number> | null;
  error_message: string | null;
  duration_ms: number | null;
  created_at: string;
  completed_at: string | null;
}

export const TEMPLATE_VARIABLES: Record<string, { label: string; placeholder: string }[]> = {
  lead_qualification: [
    { label: "Name", placeholder: "Jane Smith" },
    { label: "Company", placeholder: "Acme Corp" },
    { label: "Title", placeholder: "VP Engineering" },
    { label: "Email", placeholder: "jane@acme.com" },
    { label: "Industry", placeholder: "SaaS" },
    { label: "Company Size", placeholder: "50-200" },
    { label: "Source", placeholder: "Website form" },
    { label: "Notes", placeholder: "Requested demo" },
  ],
  email_generation: [
    { label: "Email Type", placeholder: "cold outreach" },
    { label: "Recipient Name", placeholder: "Jane Smith" },
    { label: "Recipient Title", placeholder: "VP Engineering" },
    { label: "Recipient Company", placeholder: "Acme Corp" },
    { label: "Sender Name", placeholder: "Alex" },
    { label: "Context", placeholder: "Met at conference" },
    { label: "Tone", placeholder: "professional" },
    { label: "Key Points", placeholder: "Product benefits" },
    { label: "Call to Action", placeholder: "Schedule a demo" },
  ],
  meeting_summary: [
    { label: "Meeting Title", placeholder: "Q1 Planning" },
    { label: "Meeting Date", placeholder: "2026-06-26" },
    { label: "Attendees", placeholder: "Jane, Alex, Bob" },
    { label: "Transcript", placeholder: "Paste meeting transcript..." },
  ],
  voice_crm_assistant: [
    { label: "User Message", placeholder: "Say or type your CRM request..." },
  ],
};
